### ISSUE TITLE
Adding callback for on_batch_end

### ISSUE THREAD

**Issue by smgutstein** [jim-meyer/lottery_ticket_pruner#17]
The keras_pruner_callback only calls back for on_train_end and on_epoch_begin. Shouldn't there be a callback for on_batch_end? 

The masked/pruned weights will be zeroed out at the start of each epoch and after the final epoch. But during an epoch, these weights will not be held to zero. As a result, after the initial data batch of an epoch, the mask will not be in use. Reapplying the mask at the end of each batch should ensure that the subsequent batch sees a masked net.

**Comment by jim-meyer** [comment]
IIRC the papers pruned "every j iterations". I skimmed the original paper again and I didn't see any better clarification of what exactly that meant. Does that mean "every j batches"? Or "every j epochs"?
Either way you make a good point that this implementation currently prunes every epoch, not "every j iterations" like the paper mentions.

**Comment by smgutstein** [comment]
This might just be a matter of my only starting to pick up the concepts in the papers, but I think of pruning as creating a mask that takes various a non-zero weights and sets them to zero (i.e. removing internal connections). I would want to apply the mask created by such pruning at each batch, to insure that the pruned weights don't reappear. 

I think of pruning as the creation of a new mask, and like you, am unclear as to how often one should create a new mask to prune further weights. If I read your code correctly, the callbacks call the `apply_pruning` method, which just enforces an existing pruning. Am I correct in this? 

Either way, thank you for this code. It was surprisingly readable and saved me a lot of time & work.

**Comment by jim-meyer** [comment]
Yes, the `apply_pruning` method is what sets the pruned weights to zero.
